temporal-sdk/src/main/java/io/temporal/internal/replay/WorkflowHistoryIterator.java

```diff
@@ -66,17 +66,44 @@
     nextPageToken = task.getNextPageToken();
   }
 
+  // Returns true if more history events are available.
+  // Server can return page tokens that point to empty pages.
+  // We need to verify that page is valid before returning true.
+  // Otherwise next() method would throw NoSuchElementException after hasNext() returning true.
   @Override
   public boolean hasNext() {
-    return current.hasNext() || !nextPageToken.isEmpty();
+    if (current.hasNext()) {
+      return true;
+    }
+    if (nextPageToken.isEmpty()) {
+      return false;
+    }
+
+    GetWorkflowExecutionHistoryResponse response = queryWorkflowExecutionHistory();
+
+    current = response.getHistory().getEventsList().iterator();
+    nextPageToken = response.getNextPageToken();
+
+    return current.hasNext();
   }
 
   @Override
   public HistoryEvent next() {
-    if (current.hasNext()) {
+    // If current has next element then simply return it, otherwise check if we have next page token
+    // If no page token available then we should call next() which throws NoSuchElementException.
+    if (current.hasNext() || nextPageToken.isEmpty()) {
       return current.next();
     }
 
+    GetWorkflowExecutionHistoryResponse response = queryWorkflowExecutionHistory();
```

Comment:
I don't think it is correct. As it is still going to blow up with the response contains an empty history and hasNext is not called from outside.

My proposal is to change the line 94 to:
```java
if (this.hasNext()) {
   return current.next();
}
```
and ensure that hasNext either caches non empty events in current or returns false.

Reply:
If there is no next element then next() should throw a NoSuchElementException. This is exactly what it will do.

Reply:
Simplified the code a bit, but kept behavior same.